### PR TITLE
Restore default RPATH settings & allow packages to limit to immediate deps.

### PR DESCRIPTION
**Pull request by tgamblin**
@davydden @adamjstewart 

Package authors can now set `transitive_rpaths` in their packages.

- Default is once again to use transitive `RPATH`s for the whole DAG.
   - Some packages (netcdf) NEED RPATHs for transitive deps.

- Packages can override with `transitive_rpaths=False` in `package.py`
  - Without this, some packages (dealii) will exceed OS limits when the DAG is too large

### CODE REVIEW

**Comment by tgamblin** [comment]
@davydden: can you test this out?

**Comment by davydden** [comment]
Looks good, the number of load commands is still low (169).

**Comment by tgamblin** [comment]
@adamjstewart: see if netcdf is ok for you now

**Comment by alalazo** [comment]
Sorry, it seems I am running behind the current state :smile: See my comment on #1894

**Comment by davydden** [comment]
> If every package could select whether or not to pull-in RPATH for transitive dependencies, that would make it difficult to reason about what's going on and will encourage hacks. So -1 for me

True, but in some circumstances this is the only option.

**Comment by alalazo** [comment]
> True, but in some circumstances this is the only option.

Not really, the other option is : since there's obviously a dependency that the developers didn't state, spack will require to add that. In the case discussed in #1894 it amounts to add : 
```python
depends_on('libjpeg', type='link')
```
to `netcdf` package

**Comment by davydden** [comment]
> since there's obviously a dependency that the developers didn't state, spack will require to add that. In the case discussed in #1894 it amounts to add

of course you are right. :+1: for this solution from my side.

**Comment by adamjstewart** [comment]
> Not really, the other option is : since there's obviously a dependency that the developers didn't state, spack will require to add that.

@alalazo While this is certainly a solution, the problem is that a lot of packages don't declare dependencies of dependencies. NetCDF broke because it couldn't find jpeg, szip, or openssl. I believe this change also broke ImageMagick (see #1947). I have a feeling this affects a lot of other packages as well. I feel like a lot of package authors want to minimize the number of dependencies they list. And if one of their dependencies can't be built without installing jpeg or openssl, then they assume that those packages have to be installed as well. For the record, #1740 also depends on being able to find dependencies of dependencies. Otherwise, each of those packages will have to double it's number of dependencies.

So yes, it would be easy to add these packages as dependencies. But, we would have to do it for a lot of packages, and it wouldn't be easy to figure out which need it until someone complains that it's broken. And it could break in unexpected ways. HDF5 depends on openssl, but NetCDF doesn't. So if we don't add the dependency on openssl, NetCDF will build with the system openssl and build with an HDF5 built with a different openssl.

Basically what I'm saying is that adding dependencies of dependencies is the simplest solution and breaks the fewest things. The only time it becomes a problem is with dealii on macOS Sierra (see #1874).

**Comment by adamjstewart** [comment]
> @adamjstewart: see if netcdf is ok for you now

Works again now, thanks!

**Comment by tgamblin** [comment]
@adamjstewart: I'm pretty sure this happens because packages like `netcdf` expose their link lines to dependents through configuration, which is necessary for a static link but not for a dynamic link. Other libraries like VTK and many cmake packages do this as well -- so the netcdf lib ends up getting a `DT_NEEDED` entry for `libjpeg` and you get the link dependency that way.  `netcdf` doesn't know what HDF depends on, though, and it can change by version.  Allowing transitive `RPATHs` for some packages and not others fixes that.  It also makes sure you get the right `-L`'s in the path if you do a static link (where you can't rely on DT_NEEDED and have to put everything on the link line).

I am not sure what the best thing to do is.  I'm actually curious whether @davydden can do a static build anymore if he sets `transitive_rpaths=False` (or if dealii even supports it).  We have talked about this type of stuff before with @markcmiller86 but we did not really come to an agreement about what to do or how best to handle it.

I think the answer we've got is pretty good -- `transitive_rpaths` is `True` by default, which handles dynamic and static linking, but you can turn it off if you're running into limits.  The only other options I can think of would be symlinking like @citibeth suggested, which reduces the number of search paths, but I would like to think about how to do that in a way that's consistent with Spack views and activation.  There are a lot of places where we use symlinks right now and I think they should be thought through more.

@alalazo: thoughts?

**Comment by alalazo** [comment]
@tgamblin I would put another option on the table : make it part of the logic for the build-information interface . My basic idea is that the libraries themselves know if they are built `shared` or not. Then (pseudo-code, not real implementation): 
```python
@property
def libs(self):
    if '+shared' in spec:
        return libs
    else:
        for s in self.dependencies():
            libs += s.libs
        return libs
```

That said : I understand that this is more of a theoretical problem than it is a practical one and that the solution we have works pretty well in practice.

**Comment by alalazo** [comment]
@tgamblin @adamjstewart I took a better look at the code. I think this models better what I wanted to state above : 
```python
# in Package.py
@property
def rpath_deps(self):
    if '+shared' in spec:
        return self.libs
    else:
        libs = self.libs
        for s in self.dependencies():
            libs += s.libs
        return libs

# in build_dependency.py
def get_rpath_deps(pkg):
    """Return immediate or transitive RPATHs depending on the package."""
    rpath_deps = []
    for p in pkg.spec.dependencies(deptype='link'):
        rpath_deps += p.rpath_deps
    return rpath_deps

```
